crypto/threads_win.c

```diff
@@ -37,20 +42,347 @@ typedef struct {
+    data->thread_qps[available_qp].lock = lock;
+}
+
+void ossl_rcu_write_lock(CRYPTO_RCU_LOCK lock)
+{
+    EnterCriticalSection(&lock->write_lock);
+}
+
+void ossl_rcu_write_unlock(CRYPTO_RCU_LOCK lock)
+{
+    LeaveCriticalSection(&lock->write_lock);
+}
+
+void ossl_rcu_read_unlock(CRYPTO_RCU_LOCK lock)
+{
+    struct rcu_thr_data *data = CRYPTO_THREAD_get_local(&rcu_thr_key);
+    int i;
+
+    assert(data != NULL);
+
+    for (i = 0; i < MAX_QPS; i++) {
+        if (data->thread_qps[i].lock == lock) {
+            InterlockedAdd64(&data->thread_qps[i].qp->users, -VAL_READER);
+            data->thread_qps[i].qp = NULL;
+            data->thread_qps[i].lock = NULL;
+            return;
+        }
+    }
+}
+
+static struct rcu_qp *update_qp(CRYPTO_RCU_LOCK lock)
+{
+    uint64_t new_id;
+    uint32_t current_idx;
+    uint32_t tmp;
+
+    EnterCriticalSection(&lock->alloc_lock);
+    /*
+     * we need at least one qp to be available with one
+     * left over, so that readers can start working on
+     * one that isn't yet being waited on
+     */
+    while (lock->group_count - lock->writers_alloced < 2)
+        SleepConditionVariableCS(&lock->alloc_signal, &lock->alloc_lock, INFINITE);
+
+    current_idx = lock->current_alloc_idx;
+    /* Allocate the qp */
+    lock->writers_alloced++;
+
+    /* increment the allocation index */
+    lock->current_alloc_idx =
+        (lock->current_alloc_idx + 1) % lock->group_count;
+
+    /* get and insert a new id */
+    new_id = lock->id_ctr;
+    lock->id_ctr++;
+
+    new_id = VAL_ID(new_id);
+    InterlockedAnd64(&lock->qp_group[current_idx].users, ID_MASK);
+    InterlockedAdd64(&lock->qp_group[current_idx].users, new_id);
+
+    /* update the reader index to be the prior qp */
+    tmp = lock->current_alloc_idx;
+    InterlockedExchange(&lock->reader_idx, tmp);
+
+    /* wake up any waiters */
+    WakeAllConditionVariable(&lock->alloc_signal);
+    LeaveCriticalSection(&lock->alloc_lock);
+    return (struct rcu_qp *)&lock->qp_group[current_idx];
+}
+
+static void retire_qp(CRYPTO_RCU_LOCK lock,
+                      struct rcu_qp *qp)
+{
+    EnterCriticalSection(&lock->alloc_lock);
+    lock->writers_alloced--;
+    WakeAllConditionVariable(&lock->alloc_signal);
+    LeaveCriticalSection(&lock->alloc_lock);
+}
+
+
+void ossl_synchronize_rcu(CRYPTO_RCU_LOCK lock)
+{
+    struct rcu_qp *qp;
+    uint64_t count;
+    struct rcu_cb_item *cb_items, *tmpcb;
+
+    /* before we do anything else, lets grab the cb list */
+    cb_items = InterlockedExchangePointer((void * volatile *)&lock->cb_items, NULL);
+
+    qp = update_qp(lock);
+
+    /* wait for the reader count to reach zero */
+    do {
+        count = InterlockedOr64(&qp->users, 0);
+    } while (READER_COUNT(count) != 0);
+
+    /* retire in order */
+    EnterCriticalSection(&lock->prior_lock);
+    while (lock->next_to_retire != ID_VAL(count))
+        SleepConditionVariableCS(&lock->prior_signal, &lock->prior_lock, INFINITE);
```

Comment:
The Win32 condition variable API was added only in Vista, but we have a support level of XP (0x0501) by default. This seems like a problem (it is something I already worked around in crypto/thread/arch/thread_win.c). See the comment under OPENSSL_THREADS_WINNT_LEGACY there. Perhaps this code could be refactored to use that implementation?

Reply:
I'll look at this shortly

Reply:
There appears to be a pretty close mapping to your workaround api. I'll fix it up

Reply:
refactored to use ossl api